neqo-transport/src/connection/mod.rs

```diff
@@ -2901,6 +2919,43 @@ impl Connection {
             self.decode_ack_delay(ack_delay),
             now,
         );
+
+        let path = self.paths.primary();
+        if path.borrow().is_ecn_enabled() {
+            // RFC 9000, Section 13.4.2.1:
+            //
+            // > An endpoint that receives an ACK frame with ECN counts therefore validates
+            // > the counts before using them. It performs this validation by comparing newly
+            // > received counts against those from the last successfully processed ACK frame.
+            //
+            // RFC 9000 fails to state that this is done *per packet number space*.
+            //
+            // > If an ACK frame newly acknowledges a packet that the endpoint sent with
+            // > either the ECT(0) or ECT(1) codepoint set, ECN validation fails if the
+            // > corresponding ECN counts are not present in the ACK frame.
+            //
+            // We always mark with ECT(0) - if at all - so we only need to check for that.
+            // Also, if we sent a packet with ECT(0) and get only an ACK frame (and not an
+            // ACK-ECN frame), `ecn_counts` will be all zero and the check below will fail,
+            // so no need to explicitly check for the above.
+            //
+            // > ECN validation also fails if the sum of the increase in ECT(0) and ECN-CE counts is
+            // > less than the number of newly acknowledged packets that were originally sent with an
+            // > ECT(0) marking.
+            let newly_acked = acked_packets.len() as u64;
+            let ecn_diff = diff_ecn_count(&ecn_count, &path.borrow().ecn_count(space));
+            let sum_inc = ecn_diff[IpTosEcn::Ect0] + ecn_diff[IpTosEcn::Ce];
+            if sum_inc < newly_acked {
+                qwarn!(
+                    "ACK had {} new marks, but acked {} packets, disabling ECN",
+                    sum_inc,
+                    newly_acked
+                );
+                path.borrow_mut().disable_ecn();
+            }
+            path.borrow_mut().set_ecn_count(space, ecn_count);
+        }
```

Comment:
This is a method.  Extracting this into a method would give it a name and everything.

Reply:
Moved to `Path::validate_ecn_use()`.